google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -185,47 +187,41 @@ public String toString() {
       receivedTimeMillis = clock.millisTime();
     }
 
+    private void onBoth(LinkedBlockingQueue<String> destination) {
+      acked.getAndSet(true);
+      destination.add(ackId);
+      flowController.release(1, outstandingBytes);
+      messagesWaiter.incrementPendingMessages(-1);
+      processOutstandingBatches();
+    }
+
     @Override
     public void onFailure(Throwable t) {
       logger.log(
           Level.WARNING,
           "MessageReceiver failed to processes ack ID: " + ackId + ", the message will be nacked.",
           t);
-      acked.getAndSet(true);
-      synchronized (pendingNacks) {
-        pendingNacks.add(ackId);
-      }
-      setupPendingAcksAlarm();
-      flowController.release(1, outstandingBytes);
-      messagesWaiter.incrementPendingMessages(-1);
-      processOutstandingBatches();
+      onBoth(pendingNacks);
     }
 
     @Override
     public void onSuccess(AckReply reply) {
-      acked.getAndSet(true);
+      LinkedBlockingQueue<String> destination;
       switch (reply) {
         case ACK:
-          synchronized (pendingAcks) {
-            pendingAcks.add(ackId);
-          }
+          destination = pendingAcks;
           // Record the latency rounded to the next closest integer.
           ackLatencyDistribution.record(
```

Comment:
You could honestly just do this in either case ack/nack is still a decision that could be used to determine the correct ack latency, if you want to simplify the logic

Reply:
I think it's a little weird to do this. Maybe the user adds a try-catch block and nack instead of letting the exception bubble up. Since the exception handling path doesn't record timing, I don't think this path should either.

In practice where nacks and exceptions are hopefully rare, I don't think this should matter. Do you feel strongly about this?

Reply:
No, not strongly